Wrappers/Python/cil/optimisation/utilities/sampler.py

```diff
@@ -0,0 +1,764 @@
+        self._prob = prob
+
+        if prob is None:
+            self._prob = [1/num_indices]*num_indices
+
+        if replace:
+            self._prob_weights = self._prob
+        else:
+            self._prob_weights = [1/num_indices]*num_indices
+
+        if abs(sum(self._prob_weights)-1) > 1e-6:
+            raise ValueError('The provided prob_weights must sum to one')
+
+        if any(np.array(self._prob_weights) < 0):
+            raise ValueError(
+                'The provided prob_weights must be greater than or equal to zero')
+
+        self._type = sampling_type
+        self._num_indices = num_indices
+
+        if seed is not None:
+            self._seed = seed
+        else:
+            self._seed = int(time.time())
+
+        self._generator = np.random.RandomState(self._seed)
+        
+    @property
+    def prob_weights(self):
+        return self._prob_weights
+    
+    @property
+    def num_indices(self):
+        return self._num_indices 
+
+    def next(self):
+        """ Returns and increments the sampler """
+
+        return int(self._generator.choice(self._num_indices, 1, p=self._prob, replace=self._replace))
+
+    def __next__(self):
+        return self.next()
+
+    def get_samples(self,  num_samples=20):
+        """
+        Returns the first `num_samples` as a numpy array.
+
+        num_samples: int, default=20
+            The number of samples to return. 
+
+        Example
+        -------
+        >>> sampler=Sampler.random_with_replacement(5)
+        >>> print(sampler.get_samples())
+        [2 4 2 4 1 3 2 2 1 2 4 4 2 3 2 1 0 4 2 3]
+
+        """
+        save_generator = self._generator
+        self._generator = np.random.RandomState(self._seed)
+        output = [self.next() for _ in range(num_samples)]
+        self._generator = save_generator
+        return np.array(output)
+
+
+class Sampler():
+
+    r"""
+    A class to select from a list of indices {0, 1, …, S-1}
+    The function next() outputs a single next index from the list {0,1,…,S-1} . Different orders are possible including with and without replacement. To be run again and again, depending on how many iterations.
+
+
+    Parameters
+    ----------
+    num_indices: int
+        The sampler will select from a list of indices {0, 1, …, S-1} with S=num_indices. 
+
+    sampling_type:str
+        The sampling type used. Choose from "from_function", "sequential", "custom_order", "herman_meyer", "staggered", "random_with_replacement", "random_without_replacement" and "from_function".
+
+    order: list of indices
+        The list of indices the method selects from using next. 
+
+    prob: list of floats of length num_indices that sum to 1. 
+        For random sampling with replacement, this is the probability for each index to be called by next. 
+
+    seed:int, default=None
+        Random seed for the methods that use a numpy random number generator.  If set to None, the seed will be set using the current time.  
+
```

Comment:
This class follows the factory design pattern. It is not instatiated but has 7 static methods that will return instances of 7 different samplers, which require a variety of parameters. The idea of the factory is to simplify the creation of these instances with the static methods.

The beginning of the docstring does not reflect this. The examples, on the other hand do.

Reply:
Yes- have changed this